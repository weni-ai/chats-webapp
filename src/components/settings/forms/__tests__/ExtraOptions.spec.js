import { expect, describe, it, vi, beforeEach } from 'vitest';
import { flushPromises, mount } from '@vue/test-utils';

import { createMemoryHistory, createRouter } from 'vue-router';

import FormSectorExtraOptions from '../ExtraOptions.vue';

import unnnic from '@weni/unnnic-system';

const defaultTags = [
  { uuid: 'finance', name: 'Financeiro' },
  { uuid: 'doubts', name: 'Dúvidas' },
  { uuid: 'help', name: 'Ajuda' },
];

vi.mock('@/services/api/resources/settings/sector', () => ({
  default: {
    tags: () => Promise.resolve({ results: defaultTags }),
    update: () => Promise.resolve(),
    addTag: vi.fn(),
    removeTag: vi.fn(),
  },
}));

const sectorExtraOptionsMock = {
  uuid: '1',
  can_trigger_flows: false,
  sign_messages: false,
  can_edit_custom_fields: false,
};

const routes = [{ path: '/settings', name: 'settings' }];
const router = createRouter({
  history: createMemoryHistory(),
  routes,
});

router.push = vi.fn();

function createWrapper(props = {}) {
  const wrapper = mount(FormSectorExtraOptions, {
    props: { modelValue: sectorExtraOptionsMock, isEditing: false, ...props },
    global: {
      plugins: [router],
      stubs: {
        UnnnicSwitch: true,
      },
    },
  });

  return wrapper;
}

describe('SectorExtraOptions', () => {
  let wrapper;

  beforeEach(async () => {
    wrapper = createWrapper();
    await flushPromises();
  });

  it('should render switches and switches title', async () => {
    await wrapper.setProps({ isEditing: true });
    const switchesTitle = wrapper.find('[data-testid="switchs-title"]');
    expect(switchesTitle.text()).toBe(
      wrapper.vm.$t('sector.additional_options.title'),
    );

    const switchs = wrapper.findAllComponents('[data-testid="config-switch"]');

    expect(switchs.length).toBe(3);

    expect(switchs[0].props().textRight).toContain(
      wrapper.vm.translationTriggerFlows,
    );

    expect(switchs[1].props().textRight).toContain(
      wrapper.vm.translationSignMessages,
    );

    expect(switchs[2].props().textRight).toContain(
      wrapper.vm.$t('sector.additional_options.edit_custom_fields'),
    );

    const actions = wrapper.find(
      '[data-testid="sector-extra-options-actions"]',
    );
    expect(actions.exists()).toBe(true);
  });

  it('should render tags section (without tags)', () => {
    const tagsTitle = wrapper.find('[data-testid="tags-title"]');
    expect(tagsTitle.exists()).toBe(true);
    const tagsInput = wrapper.find('[data-testid="tags-input-tag-name"]');
    expect(tagsInput.exists()).toBe(true);
    const tagsAddButton = wrapper.find('[data-testid="tags-add-tag-button"]');
    expect(tagsAddButton.exists()).toBe(true);
    const tagsSection = wrapper.find('[data-testid="tags-group-section"]');
    expect(tagsSection.exists()).toBe(false);
  });

  it('should reder tags seciton (with tags)', async () => {
    await wrapper.setData({ tags: defaultTags });

    const tagsSection = wrapper.find('[data-testid="tags-group-section"]');
    expect(tagsSection.exists()).toBe(true);
  });

  it('should emit update:modeValue on change sector', () => {
    wrapper.vm.sector = { ...sectorExtraOptionsMock };
    expect(wrapper.emitted()['update:modelValue']).toBeTruthy();
  });

  it('should get sector tags on mounted if isEditing is true', async () => {
    const getTagsSpy = vi.spyOn(FormSectorExtraOptions.methods, 'getTags');

    const wrapper = createWrapper({ isEditing: true });
    await flushPromises();
    expect(getTagsSpy).toHaveBeenCalled();
    await flushPromises();
    expect(wrapper.vm.currentTags).toEqual(defaultTags);
    expect(wrapper.vm.tags).toEqual(defaultTags);

    await getTagsSpy.mockReset();
  });

  it('should add tag on click add tag button', async () => {
    const addTagSpy = vi.spyOn(wrapper.vm, 'addTag');
    await wrapper.setData({ tagName: 'Tag 1' });
    const addTagButton = wrapper.find('[data-testid="tags-add-tag-button"]');
    await addTagButton.trigger('click');
    expect(addTagSpy).toHaveBeenCalledWith('Tag 1');
    expect(wrapper.vm.toAddTags[0].name).toEqual('Tag 1');
    expect(wrapper.vm.tags[0].name).toEqual('Tag 1');
    addTagSpy.mockClear();
  });

  it('should prevent add duplicated tag if alredy exists', async () => {
    const unnnicAlertSpy = vi.spyOn(unnnic, 'unnnicCallAlert');
    await wrapper.setData({ tags: defaultTags });

    wrapper.vm.addTag('Financeiro');
    expect(unnnicAlertSpy).toBeCalledWith({
      props: {
        text: wrapper.vm.$t('config_chats.edit_sector.tag_already_exists'),
        type: 'error',
      },
    });
  });

  it('should remove tag from tags and add to toRemoveTags', async () => {
    const removeTagSpy = vi.spyOn(wrapper.vm, 'removeTag');

    await wrapper.setData({ tags: defaultTags, toAddTags: defaultTags });
    const tagGroup = wrapper.findComponent('[data-testid="sector-tag-group"]');
    tagGroup.vm.close(defaultTags[0]);
    expect(removeTagSpy).toHaveBeenCalledWith(defaultTags[0]);
    expect(wrapper.vm.toRemoveTags[0].name).toBe('Financeiro');

    const toAddTagsNames = wrapper.vm.toAddTags.map((tag) => tag.name);
    const tagssNames = wrapper.vm.tags.map((tag) => tag.name);

    expect(toAddTagsNames.includes('Financeiro')).toBe(false);
    expect(tagssNames.includes('Financeiro')).toBe(false);
  });

  it('should save configs on click save button if isEditing is true', async () => {
    const saveSpy = vi.spyOn(wrapper.vm, 'save');
    const updateSectorTagsSpy = vi.spyOn(wrapper.vm, 'updateSectorTags');
    const updateSectorExtraConfigsSpy = vi.spyOn(
      wrapper.vm,
      'updateSectorExtraConfigs',
    );

    await wrapper.setProps({
      sector: { ...wrapper.vm.sector, can_trigger_flows: true },
      isEditing: true,
    });

    await wrapper.setData({
      currentTags: [{ uuid: 'remove', name: 'remove' }],
      toAddTags: defaultTags,
      tags: [...defaultTags, { uuid: 'remove', name: 'remove' }],
      toRemoveTags: [{ uuid: 'remove', name: 'remove' }],
    });

    const saveButton = wrapper.findComponent('[data-testid="save-button"]');

    await saveButton.trigger('click');
    await flushPromises();
    expect(saveSpy).toHaveBeenCalled();
    expect(updateSectorExtraConfigsSpy).toHaveBeenCalled();
    expect(updateSectorTagsSpy).toHaveBeenCalled();

    saveSpy.mockClear();
    updateSectorExtraConfigsSpy.mockClear();
    updateSectorTagsSpy.mockClear();
  });

  it('should show alert on error in save configs', async () => {
    const alertSpy = vi.spyOn(unnnic, 'unnnicCallAlert');
    vi.spyOn(wrapper.vm, 'updateSectorTags');
    vi.spyOn(wrapper.vm, 'updateSectorExtraConfigs').mockRejectedValue(
      new Error('Erro ao atualizar configurações extras'),
    );

    try {
      await wrapper.vm.save();
    } catch (error) {
      console.log('test error', error);
    }

    expect(alertSpy).toHaveBeenCalledWith({
      props: {
        text: wrapper.vm.$t('sector_update_error'),
        type: 'error',
      },
      seconds: 5,
    });
  });

  it('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
