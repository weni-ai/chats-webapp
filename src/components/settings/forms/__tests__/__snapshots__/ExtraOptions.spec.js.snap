// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SectorExtraOptions > Should match the snapshot 1`] = `
"<section data-v-85728205="" class="sector-extra-options-form">
  <section data-v-85728205="" class="switchs">
    <h2 data-v-85728205="" class="switchs__title" data-testid="switchs-title">Extras options</h2>
    <unnnic-switch-stub data-v-85728205="" size="medium" textleft="" textright="Triggering message templates disabled" disabled="false" modelvalue="false" usevmodel="false" data-testid="config-switch"></unnnic-switch-stub>
    <section data-v-85728205="" class="switchs__container">
      <unnnic-switch-stub data-v-85728205="" size="medium" textleft="" textright="Disabled signature" disabled="false" modelvalue="false" usevmodel="false" data-testid="config-switch"></unnnic-switch-stub>
      <div data-v-5b8c172b="" data-v-85728205="" class="unnnic-tooltip"><svg data-v-1a90efc1="" data-v-85728205="" width="40" height="40" viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg" class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-soft" data-testid="old-map-icons">
          <path d="M20 5.00049C17.0333 5.00049 14.1332 5.88022 11.6665 7.52844C9.19971 9.17666 7.27713 11.5193 6.14181 14.2602C5.0065 17.0011 4.70945 20.0171 5.28823 22.9268C5.86701 25.8366 7.29562 28.5093 9.3934 30.6071C11.4912 32.7049 14.1639 34.1335 17.0737 34.7123C19.9834 35.291 22.9994 34.994 25.7403 33.8587C28.4811 32.7234 30.8238 30.8008 32.472 28.334C34.1203 25.8673 35 22.9672 35 20.0005C34.9957 16.0236 33.414 12.2107 30.6019 9.39864C27.7897 6.58652 23.9769 5.00479 20 5.00049V5.00049ZM20.3125 11.2505C20.6833 11.2505 21.0459 11.3605 21.3542 11.5665C21.6625 11.7725 21.9029 12.0653 22.0448 12.408C22.1867 12.7506 22.2238 13.1276 22.1515 13.4913C22.0791 13.855 21.9006 14.1891 21.6383 14.4513C21.3761 14.7135 21.042 14.8921 20.6783 14.9645C20.3146 15.0368 19.9376 14.9997 19.595 14.8578C19.2524 14.7158 18.9595 14.4755 18.7535 14.1672C18.5475 13.8588 18.4375 13.4963 18.4375 13.1255C18.4375 12.6282 18.635 12.1513 18.9867 11.7997C19.3383 11.448 19.8152 11.2505 20.3125 11.2505ZM23.125 28.1255H18.125C17.7935 28.1255 17.4755 27.9938 17.2411 27.7594C17.0067 27.5249 16.875 27.207 16.875 26.8755C16.875 26.544 17.0067 26.226 17.2411 25.9916C17.4755 25.7572 17.7935 25.6255 18.125 25.6255H19.0625C19.1454 25.6255 19.2249 25.5926 19.2835 25.534C19.3421 25.4753 19.375 25.3959 19.375 25.313V19.688C19.375 19.6051 19.3421 19.5256 19.2835 19.467C19.2249 19.4084 19.1454 19.3755 19.0625 19.3755H18.125C17.7935 19.3755 17.4755 19.2438 17.2411 19.0094C17.0067 18.7749 16.875 18.457 16.875 18.1255C16.875 17.794 17.0067 17.476 17.2411 17.2416C17.4755 17.0072 17.7935 16.8755 18.125 16.8755H19.375C20.038 16.8755 20.6739 17.1389 21.1428 17.6077C21.6116 18.0766 21.875 18.7124 21.875 19.3755V25.313C21.875 25.3959 21.9079 25.4753 21.9665 25.534C22.0251 25.5926 22.1046 25.6255 22.1875 25.6255H23.125C23.4565 25.6255 23.7745 25.7572 24.0089 25.9916C24.2433 26.226 24.375 26.544 24.375 26.8755C24.375 27.207 24.2433 27.5249 24.0089 27.7594C23.7745 27.9938 23.4565 28.1255 23.125 28.1255Z" fill="#3B414D" class="primary"></path>
        </svg><span data-v-5b8c172b="" class="unnnic-tooltip-label unnnic-tooltip-label-right" style="max-width: 15rem; left: 8px; top: 0px;" data-testid="tooltip-label">Each message exchanged will have the name of the agent who is answering <br data-v-5b8c172b=""><!----></span></div>
    </section>
    <unnnic-switch-stub data-v-85728205="" size="medium" textleft="" textright="Allow agents to edit Custom Fields" disabled="false" modelvalue="false" usevmodel="false" data-testid="config-switch"></unnnic-switch-stub>
  </section>
  <section data-v-85728205="" class="tags">
    <h2 data-v-85728205="" class="tags__title" data-testid="tags-title">Add tags <div data-v-5b8c172b="" data-v-85728205="" class="unnnic-tooltip"><svg data-v-1a90efc1="" data-v-85728205="" width="40" height="40" viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg" class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-soft" data-testid="old-map-icons">
          <path d="M20 5.00049C17.0333 5.00049 14.1332 5.88022 11.6665 7.52844C9.19971 9.17666 7.27713 11.5193 6.14181 14.2602C5.0065 17.0011 4.70945 20.0171 5.28823 22.9268C5.86701 25.8366 7.29562 28.5093 9.3934 30.6071C11.4912 32.7049 14.1639 34.1335 17.0737 34.7123C19.9834 35.291 22.9994 34.994 25.7403 33.8587C28.4811 32.7234 30.8238 30.8008 32.472 28.334C34.1203 25.8673 35 22.9672 35 20.0005C34.9957 16.0236 33.414 12.2107 30.6019 9.39864C27.7897 6.58652 23.9769 5.00479 20 5.00049V5.00049ZM20.3125 11.2505C20.6833 11.2505 21.0459 11.3605 21.3542 11.5665C21.6625 11.7725 21.9029 12.0653 22.0448 12.408C22.1867 12.7506 22.2238 13.1276 22.1515 13.4913C22.0791 13.855 21.9006 14.1891 21.6383 14.4513C21.3761 14.7135 21.042 14.8921 20.6783 14.9645C20.3146 15.0368 19.9376 14.9997 19.595 14.8578C19.2524 14.7158 18.9595 14.4755 18.7535 14.1672C18.5475 13.8588 18.4375 13.4963 18.4375 13.1255C18.4375 12.6282 18.635 12.1513 18.9867 11.7997C19.3383 11.448 19.8152 11.2505 20.3125 11.2505ZM23.125 28.1255H18.125C17.7935 28.1255 17.4755 27.9938 17.2411 27.7594C17.0067 27.5249 16.875 27.207 16.875 26.8755C16.875 26.544 17.0067 26.226 17.2411 25.9916C17.4755 25.7572 17.7935 25.6255 18.125 25.6255H19.0625C19.1454 25.6255 19.2249 25.5926 19.2835 25.534C19.3421 25.4753 19.375 25.3959 19.375 25.313V19.688C19.375 19.6051 19.3421 19.5256 19.2835 19.467C19.2249 19.4084 19.1454 19.3755 19.0625 19.3755H18.125C17.7935 19.3755 17.4755 19.2438 17.2411 19.0094C17.0067 18.7749 16.875 18.457 16.875 18.1255C16.875 17.794 17.0067 17.476 17.2411 17.2416C17.4755 17.0072 17.7935 16.8755 18.125 16.8755H19.375C20.038 16.8755 20.6739 17.1389 21.1428 17.6077C21.6116 18.0766 21.875 18.7124 21.875 19.3755V25.313C21.875 25.3959 21.9079 25.4753 21.9665 25.534C22.0251 25.5926 22.1046 25.6255 22.1875 25.6255H23.125C23.4565 25.6255 23.7745 25.7572 24.0089 25.9916C24.2433 26.226 24.375 26.544 24.375 26.8755C24.375 27.207 24.2433 27.5249 24.0089 27.7594C23.7745 27.9938 23.4565 28.1255 23.125 28.1255Z" fill="#3B414D" class="primary"></path>
        </svg><span data-v-5b8c172b="" class="unnnic-tooltip-label unnnic-tooltip-label-right" style="max-width: 23rem; left: 8px; top: 0px;" data-testid="tooltip-label">Tags are used to classify service characteristics and make it possible to search and filter chats from them, define tags that are relevant to your context <br data-v-5b8c172b=""><!----></span></div>
    </h2>
    <section data-v-85728205="" class="tags-form">
      <div data-v-42852d93="" data-v-85728205="" class="unnnic-form md tags-form__input" data-testid="tags-input-tag-name">
        <p data-v-42852d93="" class="unnnic-form__label">Tag name</p>
        <div data-v-a77f9c45="" data-v-42852d93="" class="text-input size--md text-input--icon-right-size-sm tags-form__input unnnic-form-input" data-testid="tags-input-tag-name" mask=""><input data-v-bdb91cfd="" data-v-a77f9c45="" class="tags-form__input unnnic-form-input input-itself input size-md normal tags-form__input unnnic-form-input input-itself" data-testid="tags-input-tag-name" placeholder="Example: Doubts" iconrightsize="sm" hascloudycolor="false" type="text" value="">
          <!---->
          <!---->
        </div>
        <!---->
      </div><button data-v-ed84dbab="" data-v-85728205="" data-testid="tags-add-tag-button" disabled="" class="unnnic-button unnnic-button--size-large unnnic-button--secondary">
        <!---->
        <!---->
        <!----><span data-v-ed84dbab="" class="unnnic-button__label" data-testid="button-label"> Add</span>
        <!---->
      </button>
    </section>
    <!--v-if-->
  </section>
  <!--v-if-->
</section>"
`;
